File: sdks/ts/memory/src/ingest/chunk-config.test.ts
// SPDX-License-Identifier: Apache-2.0

import { describe, expect, it } from 'vitest'
import {
  DEFAULT_CHUNK_CONFIG,
  DEFAULT_MAX_TOKENS,
  DEFAULT_MIN_TOKENS,
  DEFAULT_OVERLAP_TOKENS,
  DEFAULT_SEPARATORS,
  DEFAULT_STRATEGY,
  estimateTokens,
  validateChunkConfig,
} from './chunk-config.js'
import type { ChunkConfig } from './chunk-config.js'

describe('DEFAULT_CHUNK_CONFIG', () => {
  it('has spec-mandated maxTokens of 512', () => {
    expect(DEFAULT_CHUNK_CONFIG.maxTokens).toBe(512)
    expect(DEFAULT_MAX_TOKENS).toBe(512)
  })

  it('has spec-mandated overlapTokens of 64', () => {
    expect(DEFAULT_CHUNK_CONFIG.overlapTokens).toBe(64)
    expect(DEFAULT_OVERLAP_TOKENS).toBe(64)
  })

  it('has spec-mandated minTokens of 30', () => {
    expect(DEFAULT_CHUNK_CONFIG.minTokens).toBe(30)
    expect(DEFAULT_MIN_TOKENS).toBe(30)
  })

  it('uses recursive strategy by default', () => {
    expect(DEFAULT_CHUNK_CONFIG.strategy).toBe('recursive')
    expect(DEFAULT_STRATEGY).toBe('recursive')
  })

  it('has the spec-mandated separator hierarchy', () => {
    expect(DEFAULT_CHUNK_CONFIG.separators).toEqual(['\n\n', '\n', '. ', ' ', ''])
    expect(DEFAULT_SEPARATORS).toEqual(['\n\n', '\n', '. ', ' ', ''])
  })

  it('passes its own validation', () => {
    expect(() => validateChunkConfig(DEFAULT_CHUNK_CONFIG)).not.toThrow()
  })
})

describe('validateChunkConfig', () => {
  const validBase: ChunkConfig = {
    maxTokens: 512,
    overlapTokens: 64,
    minTokens: 30,
    strategy: 'recursive',
    separators: ['\n\n', '\n', '. ', ' ', ''],
  }

  it('accepts spec defaults', () => {
    expect(() => validateChunkConfig(validBase)).not.toThrow()
  })

  it('accepts small chunks with no overlap', () => {
    expect(() =>
      validateChunkConfig({ ...validBase, maxTokens: 64, overlapTokens: 0, minTokens: 10 }),
    ).not.toThrow()
  })

  it('accepts markdown strategy', () => {
    expect(() =>
      validateChunkConfig({ ...validBase, maxTokens: 1024, overlapTokens: 128, strategy: 'markdown' }),
    ).not.toThrow()
  })

  it('accepts markdown strategy with empty separators', () => {
    expect(() =>
      validateChunkConfig({ ...validBase, strategy: 'markdown', separators: [] }),
    ).not.toThrow()
  })

  it('rejects zero maxTokens', () => {
    expect(() => validateChunkConfig({ ...validBase, maxTokens: 0 })).toThrow(
      'maxTokens must be > 0',
    )
  })

  it('rejects negative maxTokens', () => {
    expect(() => validateChunkConfig({ ...validBase, maxTokens: -1 })).toThrow(
      'maxTokens must be > 0',
    )
  })

  it('rejects overlap equal to max', () => {
    expect(() =>
      validateChunkConfig({ ...validBase, maxTokens: 100, overlapTokens: 100 }),
    ).toThrow('overlapTokens (100) must be < maxTokens (100)')
  })

  it('rejects overlap exceeding max', () => {
    expect(() =>
      validateChunkConfig({ ...validBase, maxTokens: 100, overlapTokens: 200 }),
    ).toThrow('overlapTokens (200) must be < maxTokens (100)')
  })

  it('rejects negative overlap', () => {
    expect(() => validateChunkConfig({ ...validBase, overlapTokens: -1 })).toThrow(
      'overlapTokens must be >= 0',
    )
  })

  it('rejects minTokens equal to max', () => {
    expect(() => validateChunkConfig({ ...validBase, minTokens: 512 })).toThrow(
      'minTokens (512) must be < maxTokens (512)',
    )
  })

  it('rejects negative minTokens', () => {
    expect(() => validateChunkConfig({ ...validBase, minTokens: -5 })).toThrow(
      'minTokens must be >= 0',
    )
  })

  it('rejects unknown strategy', () => {
    expect(() =>
      validateChunkConfig({ ...validBase, strategy: 'unknown' as ChunkConfig['strategy'] }),
    ).toThrow('unknown strategy')
  })

  it('rejects empty separators with recursive strategy', () => {
    expect(() => validateChunkConfig({ ...validBase, separators: [] })).toThrow(
      'separators must be non-empty',
    )
  })
})

describe('estimateTokens', () => {
  it('returns 0 for empty string', () => {
    expect(estimateTokens('')).toBe(0)
  })

  it('returns ceil(len/4) for non-empty strings', () => {
    expect(estimateTokens('a')).toBe(1)
    expect(estimateTokens('ab')).toBe(1)
    expect(estimateTokens('abc')).toBe(1)
    expect(estimateTokens('abcd')).toBe(1)
    expect(estimateTokens('abcde')).toBe(2)
  })

  it('matches conformance fixture values', () => {
    expect(estimateTokens('hello world')).toBe(3)
    expect(estimateTokens('The quick brown fox jumps over the lazy dog.')).toBe(11)
    expect(estimateTokens('a b c d e f g h')).toBe(4)
  })

  it('is monotonic in input length', () => {
    let prev = estimateTokens('')
    for (let i = 1; i <= 100; i++) {
      const cur = estimateTokens('x'.repeat(i))
      expect(cur).toBeGreaterThanOrEqual(prev)
      prev = cur
    }
  })

  it('produces same result as Go formula for boundary cases', () => {
    // Go formula: (len + 3) / 4 using integer division
    // TS formula: Math.ceil(len / 4)
    // Both should produce identical results
    const goFormula = (len: number): number => Math.trunc((len + 3) / 4)
    for (let i = 0; i <= 50; i++) {
      const tsResult = estimateTokens('x'.repeat(i))
      const goResult = i === 0 ? 0 : goFormula(i)
      expect(tsResult).toBe(goResult)
    }
  })
})

File: go/ingest/chunk_config_test.go
// SPDX-License-Identifier: Apache-2.0
package ingest

import (
	"testing"
)

func TestDefaultChunkConfig(t *testing.T) {
	cfg := DefaultChunkConfig()
	if cfg.MaxTokens != 512 {
		t.Fatalf("MaxTokens = %d, want 512", cfg.MaxTokens)
	}
	if cfg.OverlapTokens != 64 {
		t.Fatalf("OverlapTokens = %d, want 64", cfg.OverlapTokens)
	}
	if cfg.MinTokens != 30 {
		t.Fatalf("MinTokens = %d, want 30", cfg.MinTokens)
	}
	if cfg.Strategy != StrategyRecursive {
		t.Fatalf("Strategy = %q, want %q", cfg.Strategy, StrategyRecursive)
	}
	if len(cfg.Separators) != 5 {
		t.Fatalf("Separators length = %d, want 5", len(cfg.Separators))
	}
	expectedSeps := []string{"\n\n", "\n", ". ", " ", ""}
	for i, sep := range cfg.Separators {
		if sep != expectedSeps[i] {
			t.Fatalf("Separators[%d] = %q, want %q", i, sep, expectedSeps[i])
		}
	}
}

func TestDefaultChunkConfig_returnsIndependentSlice(t *testing.T) {
	a := DefaultChunkConfig()
	b := DefaultChunkConfig()
	a.Separators[0] = "MUTATED"
	if b.Separators[0] == "MUTATED" {
		t.Fatal("DefaultChunkConfig returns shared separator slice")
	}
}

func TestValidateChunkConfig(t *testing.T) {
	cases := []struct {
		name    string
		cfg     ChunkConfig
		wantErr bool
	}{
		{
			name:    "spec defaults valid",
			cfg:     DefaultChunkConfig(),
			wantErr: false,
		},
		{
			name: "small chunks no overlap valid",
			cfg: ChunkConfig{
				MaxTokens:     64,
				OverlapTokens: 0,
				MinTokens:     10,
				Strategy:      StrategyRecursive,
				Separators:    []string{"\n\n", "\n", ". ", " ", ""},
			},
			wantErr: false,
		},
		{
			name: "markdown strategy valid",
			cfg: ChunkConfig{
				MaxTokens:     1024,
				OverlapTokens: 128,
				MinTokens:     50,
				Strategy:      StrategyMarkdown,
				Separators:    []string{"\n\n", "\n", ". ", " ", ""},
			},
			wantErr: false,
		},
		{
			name: "markdown strategy empty separators valid",
			cfg: ChunkConfig{
				MaxTokens:     512,
				OverlapTokens: 64,
				MinTokens:     30,
				Strategy:      StrategyMarkdown,
				Separators:    []string{},
			},
			wantErr: false,
		},
		{
			name: "zero maxTokens rejected",
			cfg: ChunkConfig{
				MaxTokens:     0,
				OverlapTokens: 64,
				MinTokens:     30,
				Strategy:      StrategyRecursive,
				Separators:    []string{"\n\n"},
			},
			wantErr: true,
		},
		{
			name: "negative maxTokens rejected",
			cfg: ChunkConfig{
				MaxTokens:     -1,
				OverlapTokens: 64,
				MinTokens:     30,
				Strategy:      StrategyRecursive,
				Separators:    []string{"\n\n"},
			},
			wantErr: true,
		},
		{
			name: "overlap equals max rejected",
			cfg: ChunkConfig{
				MaxTokens:     100,
				OverlapTokens: 100,
				MinTokens:     30,
				Strategy:      StrategyRecursive,
				Separators:    []string{"\n\n"},
			},
			wantErr: true,
		},
		{
			name: "overlap exceeds max rejected",
			cfg: ChunkConfig{
				MaxTokens:     100,
				OverlapTokens: 200,
				MinTokens:     30,
				Strategy:      StrategyRecursive,
				Separators:    []string{"\n\n"},
			},
			wantErr: true,
		},
		{
			name: "negative overlap rejected",
			cfg: ChunkConfig{
				MaxTokens:     512,
				OverlapTokens: -1,
				MinTokens:     30,
				Strategy:      StrategyRecursive,
				Separators:    []string{"\n\n"},
			},
			wantErr: true,
		},
		{
			name: "minTokens equals max rejected",
			cfg: ChunkConfig{
				MaxTokens:     512,
				OverlapTokens: 64,
				MinTokens:     512,
				Strategy:      StrategyRecursive,
				Separators:    []string{"\n\n"},
			},
			wantErr: true,
		},
		{
			name: "negative minTokens rejected",
			cfg: ChunkConfig{
				MaxTokens:     512,
				OverlapTokens: 64,
				MinTokens:     -5,
				Strategy:      StrategyRecursive,
				Separators:    []string{"\n\n"},
			},
			wantErr: true,
		},
		{
			name: "unknown strategy rejected",
			cfg: ChunkConfig{
				MaxTokens:     512,
				OverlapTokens: 64,
				MinTokens:     30,
				Strategy:      "unknown",
				Separators:    []string{"\n\n"},
			},
			wantErr: true,
		},
		{
			name: "empty separators with recursive rejected",
			cfg: ChunkConfig{
				MaxTokens:     512,
				OverlapTokens: 64,
				MinTokens:     30,
				Strategy:      StrategyRecursive,
				Separators:    []string{},
			},
			wantErr: true,
		},
	}
	for _, tc := range cases {
		t.Run(tc.name, func(t *testing.T) {
			err := ValidateChunkConfig(tc.cfg)
			if (err != nil) != tc.wantErr {
				t.Fatalf("ValidateChunkConfig() err = %v, wantErr = %v", err, tc.wantErr)
			}
		})
	}
}

func TestEstimateTokens(t *testing.T) {
	cases := []struct {
		name     string
		input    string
		expected int
	}{
		{"empty string", "", 0},
		{"single char", "a", 1},
		{"two chars", "ab", 1},
		{"three chars", "abc", 1},
		{"four chars", "abcd", 1},
		{"five chars", "abcde", 2},
		{"hello world", "hello world", 3},
		{"sentence", "The quick brown fox jumps over the lazy dog.", 11},
		{"spaced letters", "a b c d e f g h", 4},
	}
	for _, tc := range cases {
		t.Run(tc.name, func(t *testing.T) {
			got := EstimateTokens(tc.input)
			if got != tc.expected {
				t.Fatalf("EstimateTokens(%q) = %d, want %d", tc.input, got, tc.expected)
			}
		})
	}
}

func TestEstimateTokens_monotonic(t *testing.T) {
	prev := EstimateTokens("")
	for i := 1; i <= 100; i++ {
		text := make([]byte, i)
		for j := range text {
			text[j] = 'x'
		}
		cur := EstimateTokens(string(text))
		if cur < prev {
			t.Fatalf("non-monotonic at len %d: %d < %d", i, cur, prev)
		}
		prev = cur
	}
}
